File: final_projects/ROS1_to_ROS2_Migration_Project/jawaban_project_ros1_noetic/src/turtlesim_project_cpp/src/target_publisher.cpp
#include <rclcpp/rclcpp.hpp>
#include <my_robot_msgs/msg/coordinates2_d.hpp>
#include <random>

class TargetPublisher : public rclcpp::Node
{
public:
    TargetPublisher() : Node("target_publisher")
    {
        this->declare_parameter<double>("publish_interval", 3.0);
        double publish_interval = this->get_parameter("publish_interval").as_double();

        pub_ = this->create_publisher<my_robot_msgs::msg::Coordinates2D>("target_coordinates", 10);
        timer_ = this->create_wall_timer(std::chrono::duration<double>(publish_interval), std::bind(&TargetPublisher::sendRandomCoordinates, this));
        RCLCPP_INFO(this->get_logger(), "Target publisher has been started");
    }

private:
    rclcpp::Publisher<my_robot_msgs::msg::Coordinates2D>::SharedPtr pub_;
    rclcpp::TimerBase::SharedPtr timer_;
    std::mt19937 rng_{std::random_device{}()};
    std::uniform_real_distribution<double> dist_{0.0, 11.0};

    double randomDouble()
    {
        return dist_(rng_);
    }

    void sendRandomCoordinates()
    {
        double x = randomDouble();
        double y = randomDouble();
        publishCoordinates(x, y);
    }

    void publishCoordinates(double x, double y)
    {
        auto msg = my_robot_msgs::msg::Coordinates2D();
        msg.x = x;
        msg.y = y;
        pub_->publish(msg);
    }
};

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<TargetPublisher>());
    rclcpp::shutdown();
    return 0;
}

File: INTERNSHIP/final_projects/ROS1_to_ROS2_Migration_Project/jawaban_project_ros1_noetic/src/turtlesim_project_py/scripts/target_publisher.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
import random
from my_robot_msgs.msg import Coordinates2D


class TargetPublisher(Node):
    def __init__(self):
        super().__init__('target_publisher')
        self.declare_parameter('publish_interval', 3.0)
        publish_interval = self.get_parameter('publish_interval').get_parameter_value().double_value
        self.pub_ = self.create_publisher(Coordinates2D, "target_coordinates", 10)
        self.timer_ = self.create_timer(publish_interval, self.send_random_coordinates)
        self.get_logger().info("Target publisher has been started")

    def publish_coordinates(self, x, y):
        msg = Coordinates2D()
        msg.x = x
        msg.y = y
        self.pub_.publish(msg)

    def send_random_coordinates(self):
        x = random.uniform(0.0, 11.0)
        y = random.uniform(0.0, 11.0)
        self.publish_coordinates(x, y)


def main(args=None):
    rclpy.init(args=args)
    node = TargetPublisher()
    rclpy.spin(node)
    node.destroy_node()
    rclpy.shutdown()


if __name__ == '__main__':
    main()
